### PR TITLE
PERF: reduce overhead in groupby _cython_operation

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
The `BaseGrouper._cython_operation` has lots of checking of the dtype etc, which gives quite some overhead compared to the actual aggregation function (for small data). This is a first set of small changes to cut down that overhead.

Using the `GroupManyLabels` benchmark we have 

```python
ncols = 1000
N = 1000
data = np.random.randn(N, ncols)
labels = np.random.randint(0, 100, size=N)
df = DataFrame(data)
df_am = df._as_manager('array')
```

gives

```
In [2]: %timeit df_am.groupby(labels).sum()
48.9 ms ± 1.92 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)  <-- master
39.1 ms ± 695 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  <-- PR
```

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -587,32 +590,34 @@ def _cython_operation(
...
-        elif is_integer_dtype(values):
+        elif is_integer_dtype(dtype):
             # we use iNaT for the missing value on ints
             # so pre-convert to guard this condition
             if (values == iNaT).any():
                 values = ensure_float64(values)
             else:
                 values = ensure_int_or_float(values)
-        elif is_numeric and not is_complex_dtype(values):
-            values = ensure_float64(ensure_float(values))
```
Comment (jorisvandenbossche) [inline pandas/core/groupby/ops.py:615]: The `ensure_float` was needed before to ensure that a nullable float EA was converted to the float ndarray. But now EAs already take a different code path above.

**Comment by jbrockmendel** [comment]
can you also report how the non-AM perf is affected?

**Comment by jorisvandenbossche** [comment]
> can you also report how the non-AM perf is affected?

For the benchmark, it's not really affected. The overhead I am optimizing here is only around 1% for the BlockManager case. It's only when doing those checks 1000x instead of 1x that this becomes a noticeable bottleneck (the benchmarks has 1000 columns / single block). 

I assume that for a small Series, the overhead might also be noticeable for non-AM. But trying this, and I don't really see much difference, and that's because for a single column, the dominant time is the actual factorization (which for 1000 columns is also only done once). Using the code from above:

```
In [2]: %timeit df[0].groupby(labels).sum()
384 µs ± 4.76 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  <-- master
374 µs ± 4.18 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  <-- PR
```

**Comment by jorisvandenbossche** [comment]
(some failing tests related to (unsupported) operations on categorical column that I still need to address)

**Comment by jorisvandenbossche** [comment]
This is all passing now

File: pandas/core/dtypes/missing.py [hunk 2]
```diff
@@ -556,12 +556,12 @@ def infer_fill_value(val):
     return np.nan
 
 
-def maybe_fill(arr, fill_value=np.nan):
+def maybe_fill(arr: np.ndarray):
```
Comment (jbrockmendel) [inline pandas/core/dtypes/missing.py:559]: `-> np.ndarray`?  (or just remove the return statement i guess)

File: pandas/core/groupby/ops.py [hunk 3]
```diff
@@ -500,7 +501,9 @@ def _disallow_invalid_ops(self, values: ArrayLike, how: str):
             This is either not a valid function for this dtype, or
             valid but not implemented in cython.
         """
-        dtype = values.dtype
+        if is_numeric:
+            # never an invalid op for those dtypes, so return early as fastpath
+            return
 
         if is_categorical_dtype(dtype) or is_sparse(dtype):
```
Comment (jbrockmendel) [inline pandas/core/groupby/ops.py:508]: if we're micro-optimizing, these can be replaced with isinstance(dtype, FooDtype) checks

**Comment by jbrockmendel** [comment]
LGTM ex one requested annotation, cc @jreback

File: pandas/core/groupby/ops.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/core/groupby/ops.py:508]: I was planning to do a follow-up PR with several other dtype check optimizations, so will include it there

**Review by jreback — APPROVED** [review]
lgtm, surprised this has any speedup...

once you can fix the annotation go ahead and merge

**Pull request merged**